rust/rope/src/compare.rs

```diff
@@ -0,0 +1,550 @@
+const SSID_OPTS: i32 = _SIDD_UBYTE_OPS | _SIDD_CMP_EQUAL_EACH | _SIDD_UNIT_MASK | _SIDD_NEGATIVE_POLARITY;
+
+const SSE_STRIDE: usize = 16;
+
+/// Given two 16-byte slices, returns a bitmask where the 1 bits indicate
+/// the positions of non-equal bytes.
+///
+/// The least significant bit in the mask refers to the byte in position 0;
+/// that is, you read the mask right to left.
+///
+/// # Examples
+///
+/// ```
+/// use xi_rope::compare::fast_cmpestr_mask;
+/// let one = "aaaaaaaaaaaaaaaa";
+/// let two = "aa3aaaaa9aaaEaaa";
+/// let exp = "0001000100000100";
+/// let result = format!("{:016b}", fast_cmpestr_mask(one.as_bytes(), two.as_bytes()));
+/// assert_eq!(result.as_str(), exp);
+/// ```
+///
+#[inline(always)]
+#[doc(hidden)]
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn fast_cmpestr_mask(one: &[u8], two: &[u8]) -> i32 {
+    debug_assert_eq!(one.len(), 16);
+    debug_assert_eq!(two.len(), 16);
+    unsafe {
+        let onev = _mm_loadu_si128(one.as_ptr() as *const _);
+        let twov = _mm_loadu_si128(two.as_ptr() as *const _);
+        let mask = _mm_cmpestrm(onev, 16, twov, 16, SSID_OPTS);
+        _mm_movemask_epi8(mask)
+    }
+}
+
+/// Returns the lowest `i` for which `one[i] != two[i]`, if one exists.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn ne_idx(one: &[u8], two: &[u8]) -> Option<usize> {
+    let min_len = one.len().min(two.len());
+    let mut idx = 0;
+    loop {
+        let mask: i32;
+        // if slice is less than 16 bytes we manually pad it
+        if idx + SSE_STRIDE >= min_len {
+            let mut one_buf: [u8; 16] = [0; 16];
+            let mut two_buf: [u8; 16] = [0; 16];
+            let mut temp_idx = 0;
+            for i in idx..min_len {
+                one_buf[temp_idx] = one[i];
+                two_buf[temp_idx] = two[i];
+                temp_idx += 1;
+            }
+            mask = fast_cmpestr_mask(&one_buf, &two_buf);
+        } else {
+            mask = fast_cmpestr_mask(&one[idx..idx+SSE_STRIDE], &two[idx..idx+SSE_STRIDE]);
+        }
+        let i = mask.trailing_zeros() as usize;
+        if i != 32 { return Some(idx + i); }
+        idx += SSE_STRIDE;
+        if idx >= min_len { break; }
+    }
+    None
+}
+
+/// Returns the lowest `i` such that `one[one.len()-i] != two[two.len()-i]`,
+/// if one exists.
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn ne_idx_rev(one: &[u8], two: &[u8]) -> Option<usize> {
+    let min_len = one.len().min(two.len());
+    let one = &one[one.len() - min_len..];
+    let two = &two[two.len() - min_len..];
+    debug_assert_eq!(one.len(), two.len());
+    let mut idx = min_len;
+    loop {
+        let mask: i32;
+        if idx < SSE_STRIDE {
+            let mut one_buf: [u8; 16] = [0; 16];
+            let mut two_buf: [u8; 16] = [0; 16];
+            let mut temp_idx = SSE_STRIDE - idx;
+            for i in 0..idx {
```

Comment:
I think this might be better expressed something like `one_buf[temp_idx .. SSE_STRIDE] = one[..idx]` (not sure about exact syntax).